subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -2212,11 +2196,41 @@ static ssize_t recvfrom_dtls_common(struct tls_context *ctx, void *buf,
 				    socklen_t *addrlen)
 {
 	int ret;
+	int timeout;
+	bool is_block = is_blocking(ctx->sock, flags);
 
-	ret = mbedtls_ssl_read(&ctx->ssl, buf, max_len);
-	if (ret >= 0) {
+	do {
 		size_t remaining;
 
+		ret = mbedtls_ssl_read(&ctx->ssl, buf, max_len);
+		if (ret < 0) {
+			if (ret == MBEDTLS_ERR_SSL_WANT_READ ||
+			    ret == MBEDTLS_ERR_SSL_WANT_WRITE ||
+			    ret == MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS ||
+			    ret ==  MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS) {
+				if (!is_block) {
+					return ret;
+				}
+
+				/* Blocking timeout. */
+				timeout = dtls_get_remaining_timeout(ctx);
+
+				/* Block. */
+				k_mutex_unlock(ctx->lock);
+				ret = wait_for_reason(ctx->sock, timeout, ret);
+				k_mutex_lock(ctx->lock, K_FOREVER);
```

Comment:
Just wondering that shouldn't we use condition variable instead of plain mutex as this looks like a possible race here, wdyt?
So similar fix what we did in sockets.c in commit dde03c67709a0ed4b84d4892ca05361716b75c80

Reply:
The thing is, we don't have an asynchronous event here to signal the condvar as underneath we use `poll()` to monitor the TCP/UDP socket. So I don't really see how could we make use of a condvar here.

Anyway, to my understanding releasing the mutex here should be safe - we no longer do any TLS/DTLS operations inbetween unlock/lock (that's the main reason for the rework, to make this possible), and only use `poll()` on the underlying socket - which is protected by its own mutex.